xarray/core/resample.py

```diff
@@ -0,0 +1,344 @@
+            (e.g., applying numpy ufuncs).
+        **kwargs
+            Used to call `func(ar, **kwargs)` for each array `ar`.
+
+        Returns
+        -------
+        applied : DataArray or DataArray
+            The result of splitting, applying and combining this array.
+        """
+        if shortcut:
+            grouped = self._iter_grouped_shortcut()
+        else:
+            grouped = self._iter_grouped()
+        applied = (maybe_wrap_array(arr, func(arr, **kwargs))
+                   for arr in grouped)
+        combined = self._combine(applied, shortcut=shortcut)
+
+        # If the aggregation function didn't drop the original resampling
+        # dimension, then we need to do so before we can rename the proxy
+        # dimension we used.
+        if self._dim in combined:
+            combined = combined.drop(self._dim)
+
+        if self._resample_dim in combined.dims:
+            combined = combined.rename({self._resample_dim: self._dim})
+
+        return combined
+
+    def reduce(self, func, dim=None, axis=None, shortcut=True,
+               keep_attrs=False, **kwargs):
+        """Reduce the items in this group by applying `func` along the
+        pre-defined resampling dimension.
+
+        Note that `dim` and `axis` are set by default here and are ignored
+        if passed by the user; this ensures compatibility with the existing
+        reduce interface.
+
+        Parameters
+        ----------
+        func : function
+            Function which can be called in the form
+            `func(x, axis=axis, **kwargs)` to return the result of collapsing an
+            np.ndarray over an integer valued axis.
+        keep_attrs : bool, optional
+            If True, the datasets's attributes (`attrs`) will be copied from
+            the original object to the new one.  If False (default), the new
+            object will be returned without attributes.
+        **kwargs : dict
+            Additional keyword arguments passed on to `func`.
+
+        Returns
+        -------
+        reduced : Array
+            Array with summarized data and the indicated dimension(s)
+            removed.
+        """
+
+        def reduce_array(ar):
+            return ar.reduce(func, self._dim, axis=None, keep_attrs=keep_attrs,
+                             **kwargs)
+        return self.apply(reduce_array, shortcut=shortcut)
+
+    def _interpolate(self, kind='linear'):
+        from .dataarray import DataArray
+
+        x = self._obj[self._dim].astype('float')
+        y = self._obj.values
+        axis = self._obj.get_axis_num(self._dim)
+
+        f = interp1d(x, y, kind=kind, axis=axis, bounds_error=True,
+                     assume_sorted=True)
+        new_x = self._full_index.values.astype('float')
+
+        # construct new up-sampled DataArray
+        dummy = self._obj.copy()
+        dims = dummy.dims
+        coords = dummy.coords
+        coords[self._dim] = self._full_index
```

Comment:
I think this needs to be done manually. In groupby operations we always apply things to sliced Dataset/DataArray objects, not arrays.

There is some similar logic in core/computation.py (see `build_output_coords()`).